src/dag_algo/mod.rs

```diff
@@ -295,99 +297,38 @@ pub fn layers(
     first_layer: Vec<usize>,
     index_output: bool,
 ) -> PyResult<PyObject> {
-    let mut output_indices: Vec<Vec<usize>> = Vec::new();
-    let mut output: Vec<Vec<&PyObject>> = Vec::new();
-    // Convert usize to NodeIndex
-    let mut first_layer_index: Vec<NodeIndex> = Vec::new();
-    for index in first_layer {
-        first_layer_index.push(NodeIndex::new(index));
-    }
-
-    let mut cur_layer = first_layer_index;
-    let mut next_layer: Vec<NodeIndex> = Vec::new();
-    let mut predecessor_count: HashMap<NodeIndex, usize> = HashMap::new();
-
-    let mut layer_node_data: Vec<&PyObject> = Vec::new();
-    if !index_output {
-        for layer_node in &cur_layer {
-            let node_data = match dag.graph.node_weight(*layer_node) {
-                Some(data) => data,
-                None => {
-                    return Err(InvalidNode::new_err(format!(
-                        "An index input in 'first_layer' {} is not a valid node index in the graph",
-                        layer_node.index()
-                    )))
-                }
-            };
-            layer_node_data.push(node_data);
-        }
-        output.push(layer_node_data);
-    } else {
-        for layer_node in &cur_layer {
-            if !dag.graph.contains_node(*layer_node) {
-                return Err(InvalidNode::new_err(format!(
-                    "An index input in 'first_layer' {} is not a valid node index in the graph",
-                    layer_node.index()
-                )));
-            }
+    let node_set = dag.graph.node_indices().collect::<HashSet<NodeIndex>>();
+    for layer_node in &first_layer {
+        if !node_set.contains(&NodeIndex::new(*layer_node)) {
+            return Err(InvalidNode::new_err(format!(
+                "An index input in 'first_layer' {} is not a valid node index in the graph",
+                layer_node
+            )));
```

Comment:
This ends up iterating over the graph twice for an error check. Instead you can do it like this:

```suggestion
    for layer_node in &first_layer {
        if !dag.contains_node(NodeIndex::new(*layer_node)) {
            return Err(InvalidNode::new_err(format!(
                "An index input in 'first_layer' {} is not a valid node index in the graph",
                layer_node
            )));
```